consensus/src/ledger/indexed_merkle_tree.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use anyhow::*;
+use indexmap::IndexSet;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::{merkle_tree::MerkleTree, MerkleParameters};
+use snarkvm_utilities::ToBytes;
+
+pub struct IndexedMerkleTree<P: MerkleParameters> {
+    tree: MerkleTree<P>,
+    index: IndexSet<Digest>,
+}
+
+impl<P: MerkleParameters> Clone for IndexedMerkleTree<P> {
+    fn clone(&self) -> Self {
+        let tree = self
+            .tree
+            .rebuild::<[u8; 32], _>(DummyLength(self.index.len(), PhantomData), &[])
+            .expect("failed to clone merkle tree");
+        Self {
+            tree,
+            index: self.index.clone(),
+        }
+    }
+}
+
+struct DummyLength<T>(usize, PhantomData<T>);
+
+impl<T> Iterator for DummyLength<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        panic!("attempted to iterate on dummy length iterator");
+    }
+}
+
+impl<T> ExactSizeIterator for DummyLength<T> {
+    fn len(&self) -> usize {
+        self.0
+    }
+}
+
+fn to_digest<B: ToBytes>(input: &B) -> Result<Digest> {
```

Comment:
Shouldn't `Digest` have a `from_bytes` (or something similar) method instead?

Reply:
wasnt needed as a utility. trying to avoid using the `FromBytes`/`ToBytes` structs -- why roll our own

Reply:
I don't mean as something implementing one of those traits, but just as a method on `Digest`, so it's not a free function.

Reply:
I used it as a private/local utility function, personal style I guess. Avoids exposing the function.